### PR TITLE
Import and update the previous design.md.

### PR DESCRIPTION
**Pull request by hdevalence**

### CODE REVIEW

File: design/design.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+Communication between the client code and the rest of the node should be done
+by a tower service interface. Since the `Service` trait can abstract from a
+function call to RPC, this means that it will be possible for us to isolate
+all client code to a subprocess.
+
+### Exported types
+
+- [...]
+
+`zebrad2`
```
Comment (dconnolly) [inline design/design.md]: Hm, this won't let me suggest a change but, s/zebrad2/zebrad/

File: design/design.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+  implemented in Rust.  While it may not make sense to require this in
+  every case (for instance, it probably doesn't make sense to rewrite
+  libsecp256k1 in Rust, instead of using the same upstream library as
+  Bitcoin), we should generally aim for it.
+
+* As much as reasonably possible, Zebra should minimize trust in
+  required dependencies.  Note that "minimize number of dependencies"
+  is usually a proxy for this desideratum, but is not exactly the same:
+  for instance, a collection of crates like the tokio crates are all
+  developed together and have one trust boundary.
```
Comment (dconnolly) [inline design/design.md:23]: 👏

File: design/design.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+  
+* Zebra should be well-factored internally into a collection of
+  component libraries which can be used by other applications to
+  perform Zcash-related tasks.  Implementation details of each
+  component should not leak into all other components,
```
Comment (dconnolly) [inline design/design.md]: s/,/./

File: design/design.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+### Internal Dependencies
+
+None: this the core data structure definitions.
+
+### Responsible for
+
+- definitions of commonly used data structures, e.g.,
+  - `Block`,
+  - `Transaction`,
+  - `Address`,
```
Comment (dconnolly) [inline design/design.md:51]: These live in network now.

Comment (dconnolly) [reply design/design.md:51]: (`Address` specifically.)

**Review by dconnolly — APPROVED** [review]

File: design/design.md (hunk 1 above)
Comment (hdevalence) [reply design/design.md]: Weird. I can do the edits you suggested in a co-authored-by commit instead and then push that.

File: design/design.md (hunk 4 above)
Comment (hdevalence) [reply design/design.md:51]: Hmm, I think `MetaAddr` lives in / was moved to `network`, but this `Address` was meant as in a `t-addr` or `z-addr`

Comment (dconnolly) [reply design/design.md:51]: Ah gotcha.

File: design/design.md (hunk 2 above)
Comment (hdevalence) [reply design/design.md:23]: I would like to add another para on this kind of thing but in this PR I didn't want to add anything "new" beyond updates to the design we already had

**Pull request merged**